src/ocean/math/WideUInt.d

```diff
@@ -0,0 +1,591 @@
+                internal arithmetic.
+
+        Throws:
+            WideUIntRangeException if this WideUInt value range was to be
+            exceeded
+
+    ***************************************************************************/
+
+    public void opAddAssign ( uint rhs )
+    {
+        if (add(this.payload[0], rhs))
+            enforce(.wideint_exception, this.checkAndInc(1));
+    }
+
+    unittest
+    {
+        WideUInt num = 0;
+        for (auto i = 0; i < 1000; ++i)
+            num += (1 << 31);
+        test!("==")(num.payload[0], 0);
+        test!("==")(num.payload[1], 500);
+    }
+
+    /***************************************************************************
+
+        Mutates current number in-place, dividing it by `rhs` and calculating
+        remainder.
+
+        Params:
+            rhs = number to divide by
+
+        Returns:
+            remainder of division
+
+    ***************************************************************************/
+
+    public uint divideBy ( uint rhs )
+    {
+        ulong remainder = 0;
+
+        for (ptrdiff_t idx = this.payload.length - 1; idx >= 0; --idx)
+        {
+            remainder = (remainder << 32) + this.payload[idx];
+            ulong result = remainder / rhs;
+            remainder -= rhs * result;
+            assert(result <= uint.max);
+            this.payload[idx] = cast(uint) result;
+        }
+
+        assert(remainder <= uint.max);
+        return cast(uint) remainder;
+    }
+
+    unittest
+    {
+        WideUInt num = 56 * 47;
+        test!("==")(0, num.divideBy(47));
+        test!("==")(num.payload[0], 56);
+
+        num.payload[0] = 1337;
+        num.payload[1] = 100;
+        test!("==")(1337, num.divideBy(1 << 31));
+        test!("==")(num.payload[0], 200);
+    }
+
+    /***************************************************************************
+
+        Mutates this number in place multiplying it by a regular integer
+
+        Params:
+            rhs = value to multiply by
+
+        Throws:
+            WideUIntRangeException if resulting value can't fit in current
+            WideUInt value range
+
+    ***************************************************************************/
+
+    public void multiplyBy ( uint rhs )
+    {
+        ulong overflow = 0;
+
+        for (size_t i = 0; i < this.payload.length; ++i)
+        {
+            overflow += cast(ulong)(this.payload[i]) * rhs;
+            this.payload[i] = cast(uint) (overflow & uint.max);
+            overflow >>= 32;
+        }
+
+        enforce(.wideint_exception, overflow == 0);
+    }
+
+    unittest
+    {
+        WideUInt num = 1_000_000_000;
+        num.multiplyBy(1_000_000_000);
+        test!("==")(num.toString(), "1000000000000000000");
+    }
+
+    /***************************************************************************
+
+        Returns:
+            double precision floating point representation of this number
+
```

Comment:
It will be double.infinity. How about:
```
Returns:
double precision floating point value which is nearest to this number.
Uses the same semantics as conversion from ulong to double.
```